### PR TITLE
Add more flexibility to factory types

### PR DESCRIPTION
**Pull request by christiancp100**
### What does it do?

It merges the default types for a collection type controller and the new created methods for it. 

### Why is it needed?

To have more flexibility on the factory types

### How to test it?

- Go to a content type controller and write: 
```
export default factories.createCoreController('api::toy.toy', ({ strapi }) => ({
    // here you should have autocompletion for all the collection methods: find, findOne, update,delete... and you can create 	 
   //any other methods e.g
   hello(){}
})).hello() // this should be allowed + find, update, delete or any other you declared on the config object;
```

- Go to a content type (single or collection, it should make the distinguish automatically and provide the types accordingly) service and do the same as with the controller

- Go to a content type routes directory, to the typescript file inside, then play around following this documentation: https://docs.strapi.io/dev-docs/backend-customization/routes 

Keep in mind that if you replace the uuid of the first parameter of the functions for a collectionType or a singleType, the types should update automatically
### Related issue(s)/PR(s)

Let us know if this is related to any issue/pull request

### CODE REVIEW

File: packages/core/strapi/lib/types/factories.d.ts
Comment (Convly) [inline packages/core/strapi/lib/types/factories.d.ts:45]: My only regret here is that the given config won't be based on the related controller (giving access to custom methods).
What happens if we have a custom `helloWorld(): void` handler in the controller? Can we still use its name in the only or except ?

**Review by Convly — CHANGES_REQUESTED** [review]
Overall nice improvement, it will be nice to have something accurate at this level.

However, for the moment this might create issues because SchemaUID can be `never` if no actors (`strapi`, `app`, `plugins`) are extending the `Strapi.Schemas` interface.

We'll be able to work around this behavior at ~end of the initiative when we'll have some extensions by default (made by Strapi), but for now this shouldn't be merged directly in `main`.

File: packages/core/strapi/lib/types/factories.d.ts
Comment (christiancp100) [reply packages/core/strapi/lib/types/factories.d.ts:45]: Good point, do you suggest that we extend it to allow others or that we simply remove this type and leave it as a `string[]`?

File: packages/core/strapi/lib/types/factories.d.ts
Comment (innerdvations) [inline packages/core/strapi/lib/types/factories.d.ts:64]: So, this is working and just adding to what was already there, so I wouldn't address it in this PR, but I'm a little confused why we're exporting a function from a .d.ts.

Is it because factory.js itself is a .js file and so the function definition has to be defined like this? It seems like there has to be a better way.

File: packages/core/strapi/lib/types/factories.d.ts
Comment (christiancp100) [reply packages/core/strapi/lib/types/factories.d.ts]: Tell me if it's Ok now or if you would remove the generic for the config too

File: packages/core/strapi/lib/core-api/controller/index.d.ts
Comment (christiancp100) [reply packages/core/strapi/lib/core-api/controller/index.d.ts:26]: @Convly if we don't provide this parameter in here, it would break if we put:
```
factories.createCoreController('api::home.home', () => ({
  find(ctx, next) {
    return {};
  },
}));
```

Maybe instead of force it we can put it as optional 
```suggestion
  [method: string | number | symbol]: (ctx: Context, next?: Next) => unknown;
```

File: packages/core/strapi/lib/types/factories.d.ts
Comment (Convly) [reply packages/core/strapi/lib/types/factories.d.ts:64]: It's missing a `declare` keyword to fit the .d.ts but other than that, this is correct :slightly_smiling_face:

Comment (Convly) [reply packages/core/strapi/lib/types/factories.d.ts:64]: `export declare function foo(): 'bar';`

**Review by Convly — CHANGES_REQUESTED** [review]

File: packages/core/strapi/lib/types/factories.d.ts
Comment (christiancp100) [reply packages/core/strapi/lib/types/factories.d.ts:43]: Ooops

File: packages/core/strapi/lib/types/factories.d.ts
Comment (christiancp100) [reply packages/core/strapi/lib/types/factories.d.ts:65]: Should I add it?

**Comment by Convly** [comment]
LGTM, waiting for @innerdvations last comments & approval before merging

File: packages/core/strapi/lib/types/factories.d.ts
Comment (innerdvations) [reply packages/core/strapi/lib/types/factories.d.ts:65]: Ahhh, ok, yes, that makes sense

File: packages/core/strapi/lib/core-api/controller/index.d.ts
Comment (innerdvations) [reply packages/core/strapi/lib/core-api/controller/index.d.ts:26]: `next` isn't actually a required parameter from what I can see, in fact I'm having trouble finding a controller that even uses it rather than just returning a value. Is there somewhere that it's actually required?

Comment (christiancp100) [reply packages/core/strapi/lib/core-api/controller/index.d.ts:26]: As it is a parameter, it's not going to be required, you can ignore it and pass just the context if you want, so I don't think it would matter in this case to put it as optional or not

**Review by innerdvations — COMMENTED** [review]
Should we add the `declare` before merging? (and maybe the optional `next` depending on if it's actually optional or not)

File: packages/core/strapi/lib/types/factories.d.ts
Comment (innerdvations) [inline packages/core/strapi/lib/types/factories.d.ts:43]: Sorry, last thing, this should now default `T = SingleTypeRouterConfig | CollectionTypeRouterConfig` ?

**Pull request merged**
